engine/access/subscription/base_tracker.go

```diff
@@ -0,0 +1,166 @@
+package subscription
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/engine/common/rpc"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// GetStartHeightFunc is a function type for getting the start height.
+type GetStartHeightFunc func(context.Context, flow.Identifier, uint64) (uint64, error)
+
+// StreamingData represents common streaming data configuration for access and state_stream handlers.
+type StreamingData struct {
+	MaxStreams  int32
+	StreamCount atomic.Int32
+}
+
+func NewStreamingData(maxStreams uint32) StreamingData {
+	return StreamingData{
+		MaxStreams:  int32(maxStreams),
+		StreamCount: atomic.Int32{},
+	}
+}
+
+// BaseTracker is an interface for a tracker that provides base GetStartHeight method related to both blocks and execution data tracking.
+type BaseTracker interface {
+	// GetStartHeightFromBlockID returns the start height based on the provided starting block ID.
+	GetStartHeightFromBlockID(flow.Identifier) (uint64, error)
+	// GetStartHeightFromHeight returns the start height based on the provided starting block height.
+	GetStartHeightFromHeight(uint64) (uint64, error)
+	// GetStartHeightFromLatest returns the start height based on the latest sealed block.
+	GetStartHeightFromLatest(context.Context) (uint64, error)
+}
+
+var _ BaseTracker = (*BaseTrackerImpl)(nil)
+
+// BaseTrackerImpl is an implementation of the BaseTracker interface.
+type BaseTrackerImpl struct {
+	rootBlockHeight uint64
+	state           protocol.State
+	headers         storage.Headers
+}
+
+// NewBaseTrackerImpl creates a new instance of BaseTrackerImpl.
+//
+// Parameters:
+// - rootBlockHeight: The root block height, which serves as the baseline for calculating the start height.
+// - state: The protocol state used for retrieving block information.
+// - headers: The storage headers for accessing block headers.
+//
+// Returns:
+// - *BaseTrackerImpl: A new instance of BaseTrackerImpl.
+func NewBaseTrackerImpl(
+	rootBlockHeight uint64,
+	state protocol.State,
+	headers storage.Headers,
+) *BaseTrackerImpl {
+	return &BaseTrackerImpl{
+		rootBlockHeight: rootBlockHeight,
+		state:           state,
+		headers:         headers,
+	}
+}
+
+// GetStartHeightFromBlockID returns the start height based on the provided starting block ID.
+// If the start block is the root block, skip it and begin from the next block.
+//
+// Parameters:
+// - startBlockID: The identifier of the starting block.
+//
+// Returns:
+// - uint64: The start height associated with the provided block ID.
+// - error: An error indicating any issues with retrieving the start height.
+//
+// Expected errors during normal operation:
+// - rpc.ConvertStorageError - if there is an issue retrieving the header from the storage.
```

Comment:
`rpc.ConvertStorageError` is a function, not a type. let's use this here an elsewhere
```suggestion
// - codes.NotFound - if the block was not found in storage
// - codes.Internal - for any other error
```